File: src/chess/CheckTester.java
package chess;

import javax.annotation.Nonnull;
import java.awt.*;
import java.util.ArrayList;
import java.util.Map;

/**
 * Class: CheckTester
 * Determines which player is in check (if any).
 *
 * @author Alex Hadi
 * @version May 2019
 */
public class CheckTester {
    // The current state
    @Nonnull private final ChessState chessState;

    // The positions of each player's king (index 0 maps to position for player 0's king, etc.)
    @Nonnull private final ArrayList<Point> kingPositions = new ArrayList<>();

    // The positions and pieces of each player's pieces (index 0 maps to positions/pieces for player 0, etc.)
    @Nonnull private final ArrayList<Map<Point, Piece>> playerPieces = new ArrayList<>();

    // Whether each player is in check (index 0 is true if player 0 is in check, etc.)
    @Nonnull private final ArrayList<Boolean> playersChecked = new ArrayList<>();

    /**
     * Constructor: CheckTester
     * Creates a new CheckTester object.
     *
     * @param chessState The current state of the game.
     */
    public CheckTester(@Nonnull ChessState chessState) {
        this.chessState = chessState;

        // Loops through and adds each player's pieces and king position to the corresponding instance variables.
        for (int playerId = 0; playerId < Constants.NUM_PLAYERS; playerId++) {
            playerPieces.add(chessState.getPlayerPieces(playerId));
            chessState.getKingPosition(playerId).ifPresent(kingPositions::add);
        }

        // Loops through and determines which player(s) are in check (if any)
        for (int playerId = 0; playerId < Constants.NUM_PLAYERS; playerId++) {
            playersChecked.add(playerInCheck(playerId));
        }
    }

    /**
     * Method: playerInCheck
     * Determines if the given player is in check.
     *
     * @param playerId The player ID to check for (0 or 1).
     * @return true if this player is in check (otherwise false).
     */
    private boolean playerInCheck(int playerId) {
        // Gets the position of the player's king and their enemy's pieces.
        Point myKingPosition = kingPositions.get(playerId);
        Map<Point, Piece> enemyPlayerPieces = playerPieces.get(1 - playerId);

        // Loops through and determines if any of the enemy's pieces
        // can legally attack their king (means they are in check).
        for (Map.Entry<Point, Piece> entry : enemyPlayerPieces.entrySet()) {
            Point position = entry.getKey();

            ChessRules chessRules = new ChessRules(position, chessState);
            if (chessRules.isLegalMove(myKingPosition)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Getter: getCheckedPlayer
     * Returns which player is in check (if any).
     *
     * @return The player ID of the player in check (or -1 if neither are in check).
     */
    public int getCheckedPlayer() {
        // Finds the index of "true" in the list (or -1 if both indices are false).
        return playersChecked.indexOf(true);
    }
}

File: src/chess/CheckTester.kt
package chess

import java.awt.Point
import java.util.*

/**
 * Class: CheckTester
 * Determines which player is in check (if any).
 *
 * @author Alex Hadi
 * @version May 2020
 */
class CheckTester(private val chessState: ChessState) {
    // The positions of each player's king (index 0 maps to position for player 0's king, etc.)
    private val kingPositions = ArrayList<Point>()

    // The positions and pieces of each player's pieces (index 0 maps to positions/pieces for player 0, etc.)
    private val playerPieces = ArrayList<Map<Point, Piece>>()

    // Whether each player is in check (index 0 is true if player 0 is in check, etc.)
    private val playersChecked = ArrayList<Boolean>()

    init {
        for (playerId in 0 until Constants.NUM_PLAYERS) {
            this.playerPieces.add(this.chessState.getPlayerPieces(playerId))
            this.chessState.getKingPosition(playerId)?.let(this.kingPositions::add)
        }

        // Loops through and determines which player(s) are in check (if any)
        this.run {
            if (this.kingPositions.size != 2) {
                return@run
            }

            for (playerId in 0 until Constants.NUM_PLAYERS) {
                this.playersChecked.add(this.playerInCheck(playerId))
            }
        }
    }

    /**
     * Method: playerInCheck
     * Determines if the given player is in check.
     *
     * @param playerId The player ID to check for (0 or 1).
     * @return true if this player is in check (otherwise false).
     */
    private fun playerInCheck(playerId: Int): Boolean {
        // Gets the position of the player's king and their enemy's pieces.
        val myKingPosition = this.kingPositions[playerId]
        val enemyPlayerPieces = this.playerPieces[1 - playerId]

        // Determines if any of the enemy's pieces can legally attack their king (means they are in check).
        return enemyPlayerPieces.any { ChessRules(it.key, this.chessState).isLegalMove(myKingPosition) }
    }

    /**
     * Getter: getCheckedPlayer
     * Returns which player is in check (if any).
     *
     * @return The player ID of the player in check (or -1 if neither are in check).
     */
    fun playerInCheck(): Int? {
        val checkedPlayer = this.playersChecked.indexOf(true)
        return if (checkedPlayer != -1) checkedPlayer else null
    }
}
